### PR TITLE
WRKLDS-728: Disable apiservers

### PR DESCRIPTION
**Pull request by ingvagabund**
Extend the OA init code to explicitly disable individual apiservers. The change is important for explicitly disabling DeploymentConfigs and Builds API through capabilities. The controlplane configuration on-disk data type gets rendered by OA operator in https://github.com/openshift/cluster-openshift-apiserver-operator/pull/532.

### CODE REVIEW

**Comment by dgrisonnet** [comment]
I think we will need to touch some other parts of the codebase to truly be able to enable/disable APIs. The current change only stop serving the APIs, but the underneath logic will still be executed even when the APIs are disabled.

The first one that I can see is the post start hooks: https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go#L444-L473. Some of these will not be needed if their API is disabled.

Another part that will need to be gated on the config are the various caches configurations in https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftapiserver/config.go. I don't think we should create the cache of an API that is disabled. Same goes for the controllers that are normally used to fill these cache and respond to the apiserver.

There might be some other parts of the code that will need to be tweaked, but I think we should also try to enable/disable the logic behind the APIs.

**Comment by dgrisonnet** [comment]
I like the change to have an allow list with the API that can be disabled today. That said we still have the same problem as before from https://github.com/openshift/openshift-apiserver/pull/366#issuecomment-1638153133 where the inner logic (controller, cache, ...) will still be executed even if an API is disabled no?

Also, I think it would be a great addition to add a log line and maybe even a metric telling which API is enabled and which one is disabled.

**Comment by ingvagabund** [comment]
> I like the change to have an allow list with the API that can be disabled today. That said we still have the same problem as before from https://github.com/openshift/openshift-apiserver/pull/366#issuecomment-1638153133 where the inner logic (controller, cache, ...) will still be executed even if an API is disabled no?

Checking the PostStart hooks I don't see anything related to DCs/Builds under https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go#L453-L484. There's also no DCs/Builds informer mentioned in https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go#L98-L142 or https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftapiserver/informers.go#L90-L99.

**Comment by dgrisonnet** [comment]
This looks good to me overall, but we might also want to disable these admission plugins https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftadmission/register.go#L37-L38 no?

**Comment by ingvagabund** [comment]
Those are registering handlers, which should not get invoked at all. Plus, `RegisterOpenshiftAdmissionPlugins` is invoked in `RegisterAllAdmissionPlugins` which is invoked in `init()` function. Too late to read the config and disable the handler.

**Comment by dgrisonnet** [comment]
We could maybe leverage the admission options to disable them? https://github.com/openshift/openshift-apiserver/blob/master/pkg/cmd/openshift-apiserver/openshiftapiserver/config.go#L212

But as you mentioned it should be impossible to invoke them so disabling them doesn't bring much value

File: pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go [hunk 1]
```diff
@@ -416,15 +425,59 @@ func addAPIServerOrDie(delegateAPIServer genericapiserver.DelegationTarget, apiS
...
+	// All API servers are enabled by default (nothing new to enable -> ignore the list of enabled versions)
+	for name, initFnc := range apiServerInitializers {
+		if _, exists := apiServers[name]; exists {
+			// All API servers are serving v1 resources
+			if sets.NewString(apiServers[name].DisabledVersions...).Has("v1") {
+				continue
+			}
+		}
+		delegateAPIServer = addAPIServerOrDie(delegateAPIServer, initFnc)
+	}
```
Comment (dgrisonnet) [inline pkg/cmd/openshift-apiserver/openshiftapiserver/openshift_apiserver.go:480]: nit: could you move that code inside a function to make `New` more readable?

**Review by dgrisonnet — APPROVED** [review]
/lgtm
/approve

**Pull request merged**